actors/verifreg/src/lib.rs

```diff
@@ -675,6 +685,148 @@ impl Actor {
+                let maybe_alloc = allocs.get::<Address, AllocationID, Allocation>(claim_alloc.client, claim_alloc.allocation_id)
+                .context_code(
+                    ExitCode::USR_ILLEGAL_STATE,
+                    "HAMT lookup failure getting allocation"
+                )?;
+
+                let alloc = match maybe_alloc {
+                    None => {
+                        ret_gen.add_fail(ExitCode::USR_ILLEGAL_ARGUMENT);
+                        info!(
+                            "claim references allocation id {} that does not belong to provider", claim_alloc.allocation_id,
+                        );
+                        continue;
+                    }
+                    Some(a) => a,
+                };
+   
+                if !can_claim_alloc(&claim_alloc, &alloc, rt.curr_epoch()) {
+                    ret_gen.add_fail(ExitCode::USR_FORBIDDEN);
+                    info!(
+                        "invalid sector {:?} for allocation {}", claim_alloc.sector_id, claim_alloc.allocation_id,
+                    );
+                    continue
+                }
+
+                let new_claim = Claim{
+                    provider,
+                    client: alloc.client,
+                    data: alloc.data,
+                    size: alloc.size,
+                    term_min: alloc.term_min,
+                    term_max: alloc.term_max,
+                    term_start: rt.curr_epoch(),
+                    sector: claim_alloc.sector_id,
+                };
+
+                claims.put::<Address, ClaimID, Claim>(provider, claim_alloc.allocation_id, new_claim)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, format!("failed to write claim {}", claim_alloc.allocation_id))?;
+                
+                allocs.remove::<Address, AllocationID, Allocation>(claim_alloc.client, claim_alloc.allocation_id)
+                .context_code(ExitCode::USR_ILLEGAL_STATE, format!("failed to remove allocation {}", claim_alloc.allocation_id))?;
+
+                let acc = match client_burns.get(&client_id) {
+                    Some(dc) => dc.clone(),
+                    None => DataCap::zero(),
+                };
+                client_burns.insert(client_id, acc + claim_alloc.piece_size);
+                ret_gen.add_success();
+            }
+            st.allocations = allocs.root().context_code(ExitCode::USR_ILLEGAL_STATE, "failed to flush allocation table")?;  
+            st.claims = claims.root().context_code(ExitCode::USR_ILLEGAL_STATE, "failed to flush claims table")?;  
+            Ok(())
+        })
+        .context("state transaction failed")?;
+        let st: State = rt.state()?;
+        _ = st;
+        for (client, dc_burn) in client_burns {
+            // TODO uncomment when datacap token integration lands in #514 and burn helper is implemented
+            //burn(st.token, client, dc_burn)
```

Comment:
This is simpler than you think. The datacap for each allocation is currently in the verifreg's own balance. You just need to accumulate a total and burn(self, amount).

Reply:
ah nice